app/src/main/java/ru/yandex/architectureproject/TaskViewModel.kt

```diff
@@ -0,0 +1,54 @@
+package ru.yandex.architectureproject
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import ru.yandex.architectureproject.domain.usecase.AddTaskUseCase
+import ru.yandex.architectureproject.domain.usecase.DeleteTaskUseCase
+import ru.yandex.architectureproject.domain.usecase.GetAllTasksUseCase
+import ru.yandex.architectureproject.domain.usecase.CompleteTaskUseCase
+import ru.yandex.architectureproject.domain.usecase.IncompleteTaskUseCase
+
+class TaskViewModel(
+    private val addTaskUseCase: AddTaskUseCase,
+    private val deleteTaskUseCase: DeleteTaskUseCase,
+    private val getAllTasksUseCase: GetAllTasksUseCase,
+    private val completeTaskUseCase: CompleteTaskUseCase,
+    private val incompleteTaskUseCase: IncompleteTaskUseCase
+) : ViewModel() {
+
+    private val taskForDeletionJobMap: MutableMap<Int, Job?> = mutableMapOf()
+
+    fun reduce(action: TaskAction) {
+        viewModelScope.launch {
+            when (action) {
+                is TaskAction.LoadTasks -> loadTasks()
+
+                is TaskAction.AddTask -> {
+                    addTaskUseCase(action.task)
+                    reduce(TaskAction.LoadTasks)
+                }
+
+                is TaskAction.UpdateTaskStatus -> {
+                    if (action.isDone) {
+                        taskForDeletionJobMap[action.taskId] = coroutineContext[Job]
+                        completeTaskUseCase(action.taskId)
+                    } else {
+                        taskForDeletionJobMap[action.taskId]?.cancel()
```

Comment:
Тут помимо отмены лучше джобу ещё из мапы удалить:

```suggestion
                        taskForDeletionJobMap[action.taskId]?.remove()?.cancel()
```